### ISSUE TITLE
GitHub Actions ingestion: providing validation data

### ISSUE THREAD

**Issue by dcoraboeuf** [nemerosa/ontrack#968]
In release 4.1.35, the ingestion of GH Actions workflows does not allow validation data. Design and implement a way to provide this data.

## Using actions

For actions running in the same workflow, the identity of the build to validate can rely on the current run ID:

* Getting the project & branch from the GitHub ref
* Getting the build from the run ID
* Getting the validation from the actions parameters or using the current step
* Setting the data

This can rely on a GraphQL mutation specific to GitHub ingestion, like `gitHubIngestionSetData`.

Another type of action can be used in another workflow to identify the build using other parameters:

* Actual build name
* Or actual build release property value

For example:

```yaml
- name: Another step
- name: Validation data
   uses: nemerosa/ontrack-github-ingestion-validation-data
   with:
      # Identifying the build to validate
      # By default using the current run ID

      # For another build
      # Using a build name
      # build-name: xxx
      # Using a build release property value (label)
      # build-label: yyy
      # Optionally targeting another branch or repository
      # repository: ppp
      # ref: xxx

      # Which validation stamp to use
      # By default using the current step
      # But a validation can be specified explicitly
      # validation: my-validation

      # Data to provide
      # Explicitely using validation data types in Ontrack
      # For example:
      # Test summary
      validation-data:
         type: "net.nemerosa.ontrack.extension.general.validation.TestSummaryValidationDataType"
         data:
            passed: 15
            skipped: 8
            failed: 5
      # Using shortcuts for well known types:
      test-summary-validation-data:
        passed: 15
        skipped: 8
        failed: 5
     # Or injecting a file
     validation-data-file:
        type: junit|metrics
        path: path.json|yaml|xml|properties
```

### `validation-data-file`

Only a few file types to be supported:

* `junit` - XML JUnit report, will be parsed locally and converted to test summary data
* `metrics` - properties file listing the metrics:

```properties
metric-one = 1.0
metric-two = 2.0
```

### Ontrack API

The action will:

* validate the arguments
* prepare the validation data
* call a GraphQL mutation

This mutation will look like (pseudo code):

```graphql
mutation {
   gitHubIngestionValidateDataByRunId(input: {
      repository: String,
      ref: String,
      runId: Long,
      validation: String!,
      validationData: ValidationDataInput(type: String!, data: JSON!),
   }) {
      validationRun { ... }
   }
}
```

Variants need to be provided for the identification of the build:

```
 gitHubIngestionValidateDataByBuildName(input: {
    repository: String,
    ref: String,
    buildName: String!,
    ...

 gitHubIngestionValidateDataByBuildLabelinput: {
    repository: String,
    ref: String,
    buildLabel String!,
    ...
```

### Ontrack access

How GitHub knows how to connect to Ontrack?

Assuming two things:

* `ONTRACK_URL` - an environment variable containing the Ontrack base URL
* `ONTRACK_TOKEN` - a secret containing the secret to connect to Ontrack

**Comment by dcoraboeuf** [comment]
## Using actions

For actions running in the same workflow, the identity of the build to validate can rely on the current run ID:

* Getting the project & branch from the GitHub ref
* Getting the build from the run ID
* Getting the validation from the actions parameters or using the current step
* Setting the data

This can rely on a GraphQL mutation specific to GitHub ingestion, like `gitHubIngestionSetData`.

Another type of action can be used in another workflow to identify the build using other parameters:

* Actual build name
* Or actual build release property value

For example:

```yaml
- name: Another step
- name: Validation data
   uses: nemerosa/ontrack-github-ingestion-validation-data
   with:
      # Identifying the build to validate
      # By default using the current run ID

      # For another build
      # Using a build name
      # build-name: xxx
      # Using a build release property value (label)
      # build-label: yyy
      # Optionally targeting another branch or project
      # project: ppp
      # branch: xxx

      # Which validation stamp to use
      # By default using the current step
      # But a validation can be specified explicitely
      # validation: my-validation
```

**Comment by dcoraboeuf** [comment]
The work at client side (GitHub Actions) will be done at https://github.com/nemerosa/ontrack-github-ingestion-validation-data

**Comment by dcoraboeuf** [comment]
Available in 4.2.3
